File: zcat/zcat.cpp
//This is a comment
//I love comments

#include <cassert>
#include <cstdint>
#include <iostream>
#include <fstream>
#include <vector>

class BitStream
{
    std::istream &_is;
    unsigned _bits = 0, _window = 0;
public:
    unsigned cnt = 0;
    BitStream(std::istream &is) : _is(is) { }

    int readBits(unsigned n)
    {
        for (; _bits < n; _bits += 8)
        {
            int c = _is.get();
            if (c == -1) return -1;
            _window |= c << _bits;
        }

        int ret = _window & (1 << n) - 1;
        _window >>= n, _bits -= n, cnt += n;
        return ret;
    }
};

class LZW
{
    const unsigned _maxbits;
    unsigned _oldcode, _n_bits = 9;
    char _finchar;
    std::vector<std::pair<unsigned, char>> _dict;
    std::ostream &_os;
public:
    unsigned n_bits() const { return _n_bits; }
    void reset() { _n_bits = 9; _dict.clear(); }

    LZW(unsigned maxbits, bool block_mode, char first, std::ostream &os)
      : _maxbits(maxbits), _oldcode(first), _finchar(first), _os(os)
    {
        //why did they design it this way?
        if (block_mode)
            _dict.push_back(std::pair<unsigned, char>(0, 0));
    }

    inline void code(const unsigned in)
    {
        assert(in >= 0 && in <= 65535 && in <= _dict.size() + 256);
        auto c = in;
        std::vector<char> stack;

        if (c == _dict.size() + 256)
            stack.push_back(_finchar), c = _oldcode;

        while (c >= 256U)
        {
            auto foo = _dict[c - 256];
            stack.push_back(foo.second);
            c = foo.first;
        }

        _os.put(_finchar = c);

        while (stack.size())
            _os.put(stack.back()), stack.pop_back();

        if (_dict.size() + 256 < 1U << _maxbits)
            _dict.push_back(std::pair<unsigned, char>(_oldcode, _finchar));

        if (_n_bits < _maxbits && _dict.size() + 256 > (1U << _n_bits) - 1)
            ++_n_bits;

        _oldcode = in;
    }
};

int main(int argc, char **argv)
{
    std::ostream * const os = &std::cout;
    std::istream *is = &std::cin;
    std::ifstream ifs;

    if (argc > 1)
        ifs.open(argv[1]), is = &ifs;

    BitStream bis(*is);
    assert(bis.readBits(16) == 0x9d1f);
    const unsigned maxbits = bis.readBits(5);
    assert(maxbits <= 16);
    bis.readBits(2);
    const bool block_mode = bis.readBits(1) ? true : false;
    bis.cnt = 0; //counter moet op nul om later padding te berekenen

    int first = bis.readBits(9);
    assert(first >= 0 && first < 256);
    os->put(first);
    LZW lzw(maxbits, block_mode, first, *os);

    for (int code; (code = bis.readBits(lzw.n_bits())) != -1;)
    {
        if (code == 256 && block_mode)
        {
            //other max. bits not working yet :S
            assert(maxbits == 13 || maxbits == 15 || maxbits == 16);

            //padding?!
            for (const unsigned nb3 = lzw.n_bits() << 3; (bis.cnt - 1U + nb3) % nb3 != nb3 - 1U;)
                bis.readBits(lzw.n_bits());

            lzw.reset();
        }
        else
        {        
            lzw.code(code);
        }
    }

    os->flush();
    ifs.close();
    return 0;
}

File: zcat/zcat.py
#!/usr/bin/pypy3

#Usage ./zcat.py archive.Z > archive

import sys

class BitInputStream:
    def __init__(self, file):
        self.f = open(file, "rb")
        self.bits = self.window = self.cnt = 0
    def readBits(self, n):
        while self.bits < n:
            b = self.f.read(1)
            if len(b) == 0:
                return -1
            c = int.from_bytes(b, 'little')
            self.window |= c << self.bits
            self.bits += 8
        ret = self.window & (1 << n) - 1
        self.window >>= n
        self.bits -= n
        self.cnt += n
        return ret

class LZW:
    def __init__(self, maxbits):
        self.maxbits = maxbits
        self.oldcode = 0
        self.finchar = 0
        self.dict = list()
    def code(self, incode):
        assert incode != 256 and incode <= len(self.dict) + 256
        c = incode;
        stack = bytearray()
        if c == len(self.dict) + 256:
            stack.append(self.finchar)
            c = self.oldcode
        while c >= 256:
            stack.append(self.dict[c - 256][1])
            c = self.dict[c - 256][0]
        self.finchar = c
        if len(self.dict) + 256 < (1 << self.maxbits):
            self.dict.append((self.oldcode, self.finchar))
        stack.append(self.finchar)
        stack.reverse()
        self.oldcode = incode
        return stack

if __name__ == "__main__":
    bis = BitInputStream(sys.argv[1])
    assert bis.readBits(16) == 0x9d1f
    maxbits = bis.readBits(7)
    assert maxbits >= 9 and maxbits <= 16
    assert bis.readBits(1) == 1 #block mode is always 1?
    bis.cnt = 0  #reset counter needed for cumbersome padding formula
    nbits = 9
    cnt = 0
    lzw = LZW(maxbits)
    while (c := bis.readBits(nbits)) != -1:
        cnt += 1
        if cnt == 1 << nbits - 1 and nbits != maxbits:
            nbits += 1
            cnt = 0
        if c == 256:
            #cumbersome padding formula
            #got it only working with maxbits 13, 15 and 16
            assert maxbits == 13 or maxbits == 15 or maxbits == 16
            nb3 = nbits << 3
            while (bis.cnt - 1 + nb3) % nb3 != nb3 - 1:
                bis.readBits(nbits)
            nbits = 9
            cnt = 0
            lzw.dict.clear()
        else:
            sys.stdout.buffer.write(lzw.code(c))

